### PR TITLE
fixes #4212 raise an error if selectable and flat/aliased are sent to with_polymorphic simultaenously

### LINKED ISSUE #4212: raise an error if selectable and flat/aliased are sent to with_polymorphic simultaenously
**Migrated issue, originally created by Michael Bayer ([@zzzeek](https://github.com/zzzeek))**

the docs refer to "flat=True" a lot but don't qualify this wont work with concrete inheritance and generally makes no sense if you are already passing a selectable, which should be aliased, e.g.:

```
from sqlalchemy import *
from sqlalchemy.orm import *
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.declarative import declared_attr
from sqlalchemy.ext.declarative import AbstractConcreteBase

Base = declarative_base()


class OrderItem(AbstractConcreteBase, Base):
    pass


class OrderTestItem(OrderItem):
    __tablename__ = 'order_test_item'

    order_id = Column(Integer, ForeignKey("order.id"), primary_key=True)
    test_id = Column(Integer, ForeignKey("test.id"), primary_key=True)

    test = relationship('Test')
    order = relationship('Order')

    __mapper_args__ = {
        'polymorphic_identity': 'order_test_item',
        'concrete': True
    }


class User(Base):
    __tablename__ = 'user'
    id = Column(Integer, primary_key=True)


class Test(Base):
    __tablename__ = 'test'
    id = Column(Integer, primary_key=True)


class Order(Base):
    __tablename__ = 'order'

    id = Column(Integer, primary_key=True)

    user_id = Column(Integer, ForeignKey('user.id'))

    user = relationship('User')
    items = relationship('OrderItem')

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

t1, t2 = Test(), Test()
s.add(Order(items=[OrderTestItem(test=t1), OrderTestItem(test=t2)]))
s.commit()


pjoin = polymorphic_union({
    'order_test_item': OrderTestItem.__table__,
}, 'type', 'pjoin')

order_items = with_polymorphic(
    OrderItem, [OrderTestItem], selectable=pjoin,
    flat=True  # <---- this breaks it
)


s = Session(e)
order = s.query(Order).join(Order.items.of_type(order_items)).all()
```

i dont think there are cases where both of these should be set.

### PR DESCRIPTION
**Pull request by RamonWill**
Hi, 
This is a proposed fix for Issue #4212 

When the selectable argument is passed to with_polymorphic it is already aliased. Therefore an argument error should be raised if flat is also True. I have added an if statement to the with_polymorphic function, that raises an Argument Error if selectable is none/False AND if flat=True. I have also included a test.

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [X ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

Kind Regards,

Ramon

### CODE REVIEW

File: test/orm/test_utils.py
Comment (zzzeek) [inline test/orm/test_utils.py:1173]: so this is good, but what I want to do here is make this test use a lot fewer APIs.   I guess the test here is making use of the full test case for #4212, but for testing the actual thing here we can do something much more minimal, and it can go in test/orm/inheritance/test_with_poly.py and make use of the existing fixtures:

```

diff --git a/test/orm/inheritance/test_with_poly.py b/test/orm/inheritance/test_with_poly.py
index 3ff6c191c..fc8cea0e8 100644
--- a/test/orm/inheritance/test_with_poly.py
+++ b/test/orm/inheritance/test_with_poly.py
@@ -1,4 +1,5 @@
 from sqlalchemy import and_
+from sqlalchemy import exc
 from sqlalchemy import or_
 from sqlalchemy import testing
 from sqlalchemy.orm import create_session
@@ -16,6 +17,19 @@ from ._poly_fixtures import Manager
 from ._poly_fixtures import Person
 
 
+class WithPolymorphicAPITest(_Polymorphic, _PolymorphicFixtureBase):
+    def test_no_use_flat_and_aliased(self):
+        sess = create_session()
+
+        subq = sess.query(Person).subquery()
+
+        testing.assert_raises_message(
+            exc.ArgumentError,
+            "Argument cannot be selectable and flat",
+            with_polymorphic, Person, [Engineer],
+            selectable=subq, flat=True
+        )
+
 class _WithPolymorphicBase(_PolymorphicFixtureBase):
     def test_join_base_to_sub(self):
         sess = create_session()

```

File: lib/sqlalchemy/orm/util.py
Comment (zzzeek) [inline lib/sqlalchemy/orm/util.py:937]: "the 'flat' and 'selectable' arguments cannot be passed simultaneously to with_polymorphic()"

**Review by zzzeek — CHANGES_REQUESTED** [review]
great!   A couple of changes.  let's simplify the test.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 269d387cbe40e2e69171e3aefed994eff5b0a34e of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 269d387cbe40e2e69171e3aefed994eff5b0a34e: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1849

File: test/orm/test_utils.py
Comment (RamonWill) [reply test/orm/test_utils.py]: noted, much appreciated!

**Comment by zzzeek** [comment]
this failed some tests, and the results are probably not here anymore.  if you're around, I can run them again if you want to work on this further, let me know thanks!

**Comment by RamonWill** [comment]
Hey Mike,  If you could run them again that would be appreciated.  I assumed the test were passed and it was just waiting for manual approval. Please do let me know if there's more i can do to help!
many thanks,
Ramon

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

recheck

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1849

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

pep8-recheck

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1849

**Comment by zzzeek** [comment]
OK it wants you to run the [black](https://black.readthedocs.io/en/stable/) tool on would reformat test/orm/inheritance/test_with_poly.py .   the way we normally have this stuff happen is by installing [pre-commit](https://pre-commit.com/), then go into the working directory that you checked out, and type "pre-commit install" which will add it to your current .git environment.   some background at https://www.sqlalchemy.org/develop.html#environment

**Comment by RamonWill** [comment]
Thanks, I have made the changes now.
So basically what i should've done initially was install pre-commit, which would've automatically reformatted any commits i made? If this isnt the case then i can just run the black tool on the file to have this formatting applied?

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision d925186bc40be8767194bc70f4438bb7f94c81a8 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset d925186bc40be8767194bc70f4438bb7f94c81a8 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1849

**Comment by zzzeek** [comment]
yes we use pre-commit which does all the formatting and a flake8 test before the commit goes through. 

certainly we didnt start out using all these tools but after 15 years of mostly not having them, I can safely say they save a lot of time, and the consistent formatting thing is so that we have the best chance possible of code merges going in cleanly,  so that we don't have commits that include large amounts of unrelated reformatting, and that we can have the reformatting macros right in our editors withotu worrying about them creating large reformatting artifacts in commits.

**Comment by RamonWill** [comment]
Okay noted and understood, thanks for the explanation

**Comment by RamonWill** [comment]
Hi Mike
Just a quick reminder that the changes have been made using black and I believe they have passed all the gerrit reviews this time around.
Thanks

Ramon

**Comment by zzzeek** [comment]
yup, I"m in this place where I have 3+ people all sending reviews in all day so I spend part of my day trying to get to some of them :)

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

Code-Review+2

so this looks good, do you want to try to do the changelog file?  if you were doing more issues it would be helpful if you knew how to do these.  there's a readme at https://github.com/sqlalchemy/sqlalchemy/blob/master/doc/build/changelog/README.txt

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1849

**Comment by RamonWill** [comment]
Thanks Mike, 

I have now added the changelog.

Kind Regards,
Ramon

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision eae20cdadac82ca1e0352000fe0211e20b590e8b of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset eae20cdadac82ca1e0352000fe0211e20b590e8b added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1849

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

Code-Review+2 Workflow+2

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1849

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1849 has been **merged**. Congratulations! :)

**Pull request closed without merging**
